src/main/java/org/terasology/magicalstats/system/ManaAuthoritySystem.java

```diff
@@ -0,0 +1,209 @@
+import org.terasology.entitySystem.entity.EntityManager;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.entitySystem.event.ReceiveEvent;
+import org.terasology.entitySystem.prefab.Prefab;
+import org.terasology.entitySystem.systems.BaseComponentSystem;
+import org.terasology.entitySystem.systems.RegisterMode;
+import org.terasology.entitySystem.systems.RegisterSystem;
+import org.terasology.entitySystem.systems.UpdateSubscriberSystem;
+import org.terasology.logic.characters.CharacterMovementComponent;
+import org.terasology.logic.characters.events.AttackEvent;
+import org.terasology.logic.characters.MovementMode;
+import org.terasology.logic.health.EngineDamageTypes;
+import org.terasology.logic.inventory.ItemComponent;
+import org.terasology.magicalstats.component.MagicalStatsComponent;
+import org.terasology.magicalstats.component.ManaComponent;
+import org.terasology.magicalstats.event.BeforeDrainedEvent;
+import org.terasology.magicalstats.event.BeforeManaRefillEvent;
+import org.terasology.magicalstats.event.BeforeManaRegenEvent;
+import org.terasology.magicalstats.event.DoDrainEvent;
+import org.terasology.magicalstats.event.DoManaRefillEvent;
+import org.terasology.magicalstats.event.DoManaRegenEvent;
+import org.terasology.magicalstats.event.FullManaEvent;
+import org.terasology.magicalstats.event.ManaChangedEvent;
+import org.terasology.magicalstats.event.OnDrainedEvent;
+import org.terasology.magicalstats.event.OnManaRefillEvent;
+import org.terasology.magicalstats.event.OnManaRegenEvent;
+import org.terasology.math.TeraMath;
+import org.terasology.protobuf.EntityData;
+import org.terasology.registry.In;
+import org.terasology.utilities.random.FastRandom;
+import org.terasology.utilities.random.Random;
+
+
+
+@RegisterSystem(RegisterMode.AUTHORITY)
+public class ManaAuthoritySystem extends BaseComponentSystem implements UpdateSubscriberSystem {
+    @In
+    private EntityManager entityManager;
+
+    @In
+    private org.terasology.engine.Time time;
+
+    @Override
+    public void update(float delta) {
+        for (EntityRef entity : entityManager.getEntitiesWith(ManaComponent.class)) {
+            ManaComponent mana = entity.getComponent(ManaComponent.class);
+            if (mana.currentMana <= 0) {
+                continue;
+            }
+            if (mana.currentMana == mana.maxMana || mana.manaRegenRate == 0) {
+                continue;
+            }
+
+            int manaRegenAmount = 0;
+            manaRegenAmount = manaRegen(mana, manaRegenAmount);
+
+            checkRefilled(entity, mana, manaRegenAmount);
+        }
+    }
+
+    /**
+     * Override the default behavior for an attack
+     */
+    @ReceiveEvent(components = ManaComponent.class, netFilter = RegisterMode.AUTHORITY)
+    public void onDrainEntity(AttackEvent event, EntityRef targetEntity) {
+        drainEntity(event, targetEntity);
+    }
+
+    static void drainEntity(AttackEvent event, EntityRef targetEntity) {
```

Comment:
`static` ? Default level access? Bit puzzling, although I know sometimes IntelliJ recommends weird access levels (sometimes we nod and smile at IntelliJ and pass by some of the weirdo recommendations).

Anyway this method (or `onDrainEntity` could probably use some javadoc / comments on exactly what is going on. Does *any* attack also drain mana by default? And by consuming the event while replacing it with a `DoDrainEvent` would that prevent regular combat damage, like one player hitting another on a server?